### ISSUE TITLE
Sentence tokenization and detokenization

### ISSUE THREAD

**Issue by BLKSerene** [Esukhia/pybo#37]
Since I don't speak tibetan, I'm not sure whether pybo can do sentence tokenization and detokenization for Tibetan text or not. It seems to me that tibetan sentences are separated by a whitespace, so is it okay to just split sentences by whitespace (without resorting to machine learning approaches for real sentence boundary detection)?

The trailing whitespace is preserved in the tokenized text, but after I've removed the trailing whitespace in the output, is it possible to detokenize the list of token back to a string of text (with whitespace added back to between sentences)?

**Comment by drupchen** [comment]
Short answer: I am not aware yet of any effective strategy to split Tibetan text into sentences. 

Long answer: Tibetan language uses the same strategy as Sanskrit. spaces and vertical bars only mean that a chunk of meaning is ending, but that chunk could be a full sentence or the end of a part of a sentence. In passing, spaces really mean nothing in Tibetan. They are more of an ornamentation to the text.

Terminative casual particles that are directly followed by punctuation (། chars) show with certainty that a sentence has ended, but sentences don't necessarily end with this casual particle. It could be used to make smaller chunks in a long text. There are also other hints that could be put together as heuristics to identify sentence boundaries, but I am not aware of any project successfully doing it.
(terminative particles unambiguously show the end of a sentence, but that sentence could be embedded in a longer sentence, hence the use of punctuation to filter as much as possible embedded sentences)

**Comment by mikkokotila** [comment]
To add, intuitively the concept of sentence is quite different in Tibetan as opposed to in English. In layman terms, it seems fair to say that in English language individual sentences have a greater degree of independence in respect to preceding and following sentences in comparison to Tibetan. So treating sentences in isolation from other sentences may not make practical sense in Tibetan as it may make in English. I could be wrong, but that's the feeling I get with my limited understanding. @drupchen would be able to say if I'm wrong.

In case I'm right, it does not necessarily constitute a limitation for research interest, but meaningful to keep in mind.

**Comment by drupchen** [comment]
@mikkokotila you are right: Tibetans are unaware that their language has a structure while the study of how sentences are created is studied since the Middle Age for Latin languages.
The concept of a sentence is very new to Tibetans, so they don't have any clear-cut definition.
I would say that the unit that is apprehended for Tibetan is the paragraph. That would correspond to what you are saying.

Now going astray of what can readily be implemented in NLP, I am tending more and more towards identifying sentences by a group of words that end with a verbal clause. These can be followed by particles that still are part of the sentence.

 Still, the purpose of sentence tokenization is to have chunks of text that are munchable by the next steps in NLP processing, such as Noun Phrase Identification, Shallow Parsing, or even more accurate POS Tagging. This would require to spend the time needed to come up with the heuristics I was talking about earlier. At some point, someone will have to do that job....

**Comment by mikkokotila** [comment]
I think defining the sentence boundary as an inference from a verbal clause sounds like a good place to open the door for more later. That will already unlock experimental research that focus on sentences.

**Comment by drupchen** [comment]
I'm coming back to you, @BLKSerene  and @mikkokotila about the sentence tokenization.

I have tried a very naive approach that is giving me reasonable results on prose. I didn't test it out on verses, but it should at least be an easy way to get smaller chunks out a text.

The code is [sentencify()](https://github.com/buda-base/bonlp-datasets/blob/master/prepare_datasets.py#L51). This function could have been much simpler, but it is made to be inserted in a pipeline like the one [here](https://github.com/buda-base/bonlp-datasets/blob/master/prepare_datasets.py#L96-L100), so it takes a pybo's token list as input. Its output is explained [here](https://github.com/buda-base/bonlp-datasets/blob/master/prepare_datasets.py#L62).

The idea is simply to use the terminal case particles followed by punctuation, then the particles that I call "clause boundary markers" to know where a sentence ends.

As you see, I have not yet implemented my initial idea using verbs, but this already gives pretty good results: I am able to split a whole page of text ([this string](https://github.com/buda-base/bonlp-datasets/blob/master/prepare_datasets.py#L102-L132)) into 63 sentences. They range from 4 words to about 50 words. 

Have a look at the results [here](https://github.com/buda-base/bonlp-datasets/blob/master/lighttag/test.json).

**Comment by BLKSerene** [comment]
@drupchen Thank! That's great. I'll take a look.
